sumcheck/src/multilinear.rs

```diff
@@ -0,0 +1,368 @@
+    }
+
+    /// Evaluate the multi-linear at some query $(r_0, ..., r_{{\nu} - 1}) \in \mathbb{F}^{\nu}$.
+    ///
+    /// It first computes the evaluations of the Lagrange basis polynomials over the interpolating
+    /// set ${0 , 1}^{\nu}$ at $(r_0, ..., r_{{\nu} - 1})$ i.e., the Lagrange kernel at $(r_0, ..., r_{{\nu} - 1})$.
+    /// The evaluation then is the inner product, indexed by ${0 , 1}^{\nu}$, of the vector of
+    /// evaluations times the Lagrange kernel.
+    pub fn evaluate(&self, query: &[E]) -> E {
+        let tensored_query = compute_lagrange_basis_evals_at(query);
+        inner_product(&self.evaluations, &tensored_query)
+    }
+
+    /// Similar to [`Self::evaluate`], except that the query was already turned into the Lagrange
+    /// kernel (i.e. the [`lagrange_ker::EqFunction`] evaluated at every point in the set
+    /// `${0 , 1}^{\nu}$`).
+    ///
+    /// This is more efficient than [`Self::evaluate`] when multiple different [`MultiLinearPoly`]
+    /// need to be evaluated at the same query point.
+    pub fn evaluate_with_lagrange_kernel(&self, lagrange_kernel: &[E]) -> E {
+        inner_product(&self.evaluations, lagrange_kernel)
+    }
+
+    /// Computes $f(r_0, y_1, ..., y_{{\nu} - 1})$ using the linear interpolation formula
+    /// $(1 - r_0) * f(0, y_1, ..., y_{{\nu} - 1}) + r_0 * f(1, y_1, ..., y_{{\nu} - 1})$ and assigns
+    /// the resulting multi-linear, defined over a domain of half the size, to `self`.
+    pub fn bind_least_significant_variable(&mut self, round_challenge: E) {
+        let num_evals = self.evaluations.len() >> 1;
+        #[cfg(not(feature = "concurrent"))]
+        {
+            for i in 0..num_evals {
+                self.evaluations[i] = self.evaluations[i << 1]
+                    + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+            }
+            self.evaluations.truncate(num_evals)
+        }
+
+        #[cfg(feature = "concurrent")]
+        {
+            let mut result = unsafe { utils::uninit_vector(num_evals) };
+            result.par_iter_mut().enumerate().for_each(|(i, ev)| {
+                *ev = self.evaluations[i << 1]
+                    + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1])
+            });
+            self.evaluations = result
+        }
+    }
+
+    /// Given the multilinear polynomial $f(y_0, y_1, ..., y_{{\nu} - 1})$, returns two polynomials:
+    /// $f(0, y_1, ..., y_{{\nu} - 1})$ and $f(1, y_1, ..., y_{{\nu} - 1})$.
+    pub fn project_least_significant_variable(&self) -> (Self, Self) {
+        let mut p0 = Vec::with_capacity(self.num_evaluations() / 2);
+        let mut p1 = Vec::with_capacity(self.num_evaluations() / 2);
+        for chunk in self.evaluations.chunks_exact(2) {
+            p0.push(chunk[0]);
+            p1.push(chunk[1]);
+        }
+
+        (MultiLinearPoly::from_evaluations(p0), MultiLinearPoly::from_evaluations(p1))
+    }
```

Comment:
We could try rewriting this function as:

```rs
    pub fn project_least_significant_variable(mut self) -> (Self, Self) {
        let odds: Vec<E> = self
            .evaluations
            .iter()
            .enumerate()
            .filter_map(|(idx, x)| if idx % 2 == 1 { Some(*x) } else { None })
            .collect();

        // Builds the evens multilinear from the current `self.evaluations` buffer, which saves an
        // allocation.
        let evens = {
            let evens_size = self.num_evaluations() / 2;
            for write_idx in 0..evens_size {
                let read_idx = write_idx * 2;
                self.evaluations[write_idx] = self.evaluations[read_idx];
            }
            self.evaluations.truncate(evens_size);

            self.evaluations
        };

        (Self::from_evaluations(evens), Self::from_evaluations(odds))
    }
```

This version saves an allocation compared to the current version. I'd be curious how this affects the sum-check benchmark (after we moved the `project_least_significant_variable()` call into the sum-check, as mentioned in another comment).

Reply:
Done, thank you for the suggestion!
I haven't noticed any noticeable difference on machine (x86) but maybe it is different on other archs.